Add Steam ID implementation

ObsidianMinor opened the pull request:
Closes #1

ObsidianMinor committed:
Add initial impl for SteamId

ObsidianMinor committed:
Add tests and fix issues discovered while writing tests

JurjenBiewenga commented on src/Steam.Common/SteamId.cs:
```diff
@@ -0,0 +1,492 @@
+        /// <summary>
+        /// Returns whether this is a login ID to be filled in
+        /// </summary>
+        public bool IsBlankAnonymousAccount => AccountId == 0 && IsAnonymousAccount && AccountInstance == 0;
```
I'd like to see one white line between properties to make the distinction more clear.

JurjenBiewenga commented on src/Steam.Common/SteamId.cs:
```diff
@@ -0,0 +1,492 @@
+    /// </summary>
+    public readonly struct SteamId : IEquatable<SteamId>
+    {
+        private const uint AccountIdMask = 0xFFFFFFFF;
```
We should consider whether we want to add documentation to private members as well.

ObsidianMinor replied:
I don't think it's necessary, unless the private member is incredibly vague we may add a comment but I don't think we should add too many docs.

JurjenBiewenga commented on src/Steam.Common/SteamId.cs:
```diff
@@ -0,0 +1,492 @@
+
+        private const int ChatAccountInstanceMask = 0xFFF;
+
+        private readonly static Regex Steam2Regex = new Regex(@"STEAM_([0-4]):(0|1):(\d+)");
```
The order of these modifiers is inconsistent with the rest. We should pick one way and stick with it.

ObsidianMinor replied:
I'd say `static` should be followed by `readonly`. I can't figure out any way to enforce this however.

JurjenBiewenga commented on src/Steam.Common/AccountType.cs:
```diff
@@ -0,0 +1,53 @@
+﻿namespace Steam
```
Shouldn't this and the other namespaces be `Steam.Common`?

ObsidianMinor replied:
Was your stuff in Steam.Common?

JurjenBiewenga replied:
Yeah it is, feel free to change it.

ObsidianMinor replied:
Nope, I checked, it's in `Steam` (I set the root namespace ahead of time).

JurjenBiewenga commented on tests/Steam.Common.Tests/SteamIdTests.cs:
```diff
@@ -0,0 +1,152 @@
+
+namespace Steam.Tests
+{
+    public class SteamIdTests
```
I'd like to see more tests to check the correctness of equality operators and preferably also the properties to make sure we can't accidentally break these.

ObsidianMinor replied:
I believe the "parts" tests test these well.

JurjenBiewenga commented on src/Steam.Common/SteamId.cs:
```diff
@@ -0,0 +1,492 @@
+    /// <summary>
+    /// A globally unique identifier for Steam accounts, groups, lobbies, and chat rooms
+    /// </summary>
+    public readonly struct SteamId : IEquatable<SteamId>
```
I don't think it is a bad thing to link to the Steam API for more information about this construct, if we want to do this we should keep it consistent in the whole project (Currently app, depot, and package do have this)

ObsidianMinor replied:
The Steam API docs are those docs
![image](https://user-images.githubusercontent.com/10121870/58666073-ff1af300-82f7-11e9-8ce1-1651e2e32775.png)

JurjenBiewenga replied:
That's all there is that's documented on their site? If it is then that it is fine like this

JurjenBiewenga commented on src/Steam.Common/SteamId.cs:
```diff
@@ -0,0 +1,492 @@
+            if (!match.Success)
+                throw new ArgumentException("The input string does not match the Steam3 pattern");
+
+            var typeChar = match.Groups[1].Value[0];
```
Nice use of groupings, however using named groups might make the code more readable

ObsidianMinor replied:
With constant ints?

JurjenBiewenga replied:
Nah, you can add names to groups in regex. Lke this https://docs.microsoft.com/en-us/dotnet/standard/base-types/grouping-constructs-in-regular-expressions#named-matched-subexpressions

JurjenBiewenga reviewed (changes requested):
Looks pretty good, mostly some stylistic issues. Unfortunately, I haven't been able to fact check your implementation against the steam api.

ObsidianMinor committed:
Changes for review comments:
 * Add names to regex groups
 * Add whitespace between properties and methods
 * Add check to make sure two of the same whole value are equal

ObsidianMinor committed:
Change operators on SteamId to implicit and remove Value from public API

ObsidianMinor committed:
Many fixups for SteamId
Added summaries for the rest of SteamId APIs
Rewrote tests
Added TryFromSteam2 and TryFromSteam3 APIs
Fixed a few summaries
